PR#0001: Реализован логгер с использованием паттерна Strategy

albrettto opened the pull request.

albrettto committed:
feat(LogStrategy): Реализована логика Логгера

Реализована логика Логгера с использованием паттернов Singleton и Strategy

albrettto committed:
feat(main): Добавлен main файл с примерами реализации

albrettto committed:
feat(test): Добавлен файл с тестами

albrettto committed:
chore(files): Исправлено именование файлов

albrettto committed:
feat(CI): Добавлен файл сборки и тестирования

albrettto committed:
fix(log_strategy): Исправлена ошибка с кавычками

vladushka2000 commented on log_strategy.py:
```diff
@@ -0,0 +1,64 @@
+import os
+
+
+class LogStrategy(ABC):
```
1) Слишком раздутый интерфейс, нарушен isp. Выдели основные методы, а остальные сделай приватными
2) Не нравится название класса. Именуем как сущность, избавляемся от паттернов в названии. Предлагаю Writer

vladushka2000 commented on log_strategy.py:
```diff
@@ -0,0 +1,64 @@
+from abc import ABC, abstractmethod
```
Нет ни одной докстринги. Написать для всех классов и каждого метода

vladushka2000 commented on logger.py:
```diff
@@ -0,0 +1,23 @@
+from singleton import Singleton
+import threading
+
+class Logger(Singleton):
```
Нету интерфейса для логгера

vladushka2000 commented on logger.py:
```diff
@@ -0,0 +1,23 @@
+    def set_log_strategy(self, log_strategy: LogStrategy):
+        self.log_strategy = log_strategy
+
+    def set_file_path(self, file_path: str):
```
у тебя этот метод будет иметь смысл только тогда, когда мы хотим писать в файл. Подумай, куда убрать создание пути к файлу, чтобы в консольном логгере мы не могли обратиться к этому методу

Rognak commented on .github/workflows/python-package.yml:
```diff
@@ -0,0 +1,42 @@
+
+on:
+  push:
+    branches: [ "feature" ]
```
Зачем этот фильтр?

albrettto replied:
Чтобы запуск происходил при пуше новых коммитов в эту ветку

Rognak replied:
Ты тем самым ограничиваешь себя. А что если мы продолжим разработку? Будут другие ветки и т.д.

albrettto replied:
Ну в таком случае, лучше убрать. Я уберу тогда

Rognak commented on test.py:
```diff
@@ -0,0 +1,66 @@
+
+    def setUp(self):
+        """Подготовка перед каждым тестом."""
+        self.file_path = tempfile.TemporaryDirectory()  # Используем временную директорию
```
Как вариант. Ну лучше вообще не писать в тестах в файлы

Rognak commented on test.py:
```diff
@@ -0,0 +1,66 @@
+    def test_console_log_multithreaded(self):
+        """Тест многопоточного логгирования в консоль."""
+        logger = Logger(ConsoleLogStrategy())
+        self.run_multithreaded_logging(logger, f"[INFO] Application started successfully 0")
```
Ты здесь всё привязал к многопоточке, зачем? Это можно вынести в отдельный тест кейс

Rognak reviewed (changes requested):
Необходимо исправить замечания

albrettto committed:
feat(logger): Реализован логгер

Логгер реализован с использованием паттерна Одиночка и подготовлен под реализацию паттерна Стратегия

albrettto committed:
feat(enum): Добавлен файл с enum для уровней логгирования

albrettto committed:
feat(timestamp_provider): Добавлен файл с классом для предоставления временной метки

albrettto committed:
feat(file_path_manager): Добавлен интерфейс для управления путями к файлам логов

albrettto committed:
feat(writers): Добавлены различные стратегии реализации логгера

albrettto committed:
feat(main): Добавлен основной файл проекта

В данном файле демонстрируется работа разных стратегий логгера: вывод сообщений в консоль, вывод сообщений в файл, вывод сообщений в файл КАПСОМ и смена пути к файлу для логгеров, которые записывают логи в файл

albrettto committed:
Merge remote-tracking branch 'origin/feature' into feature

# Conflicts:
#	logger.py
#	main.py
#	singleton.py

albrettto committed:
feat(CI): Добавлен файл сборки и тестирования

albrettto committed:
devops(CI): Добавлен файл сборки и тестирования

albrettto committed:
test(test): Добавлен файл с тестами основного функционала

albrettto committed:
feat(test): Добавлен файл с тестами основного функционала

albrettto commented:
Исправил замечания

albrettto committed:
chore(log_strategy): Удален старый файл логгера

albrettto committed:
chore(log_strategy): Удален старый файл логгера

albrettto committed:
Merge remote-tracking branch 'origin/feature' into feature

vladushka2000 commented on enums/level_enum.py:
```diff
@@ -0,0 +1,20 @@
+        FATAL (5): Критические ошибки.
+    """
+
+    TRACE = 1
```
зачем? Значение должно соответствовать названию. Ты потом должен оперировать полем value из LogLevel, а не name. В этом и смысл Enum - мы задаем поля, которые потом используем в коде, а значения этих полей мы можем подменять как нам удобно

albrettto replied:
Исправил

vladushka2000 commented on interfaces/ilogger.py:
```diff
@@ -0,0 +1,16 @@
+    """Интерфейс для логгера, определяющий основные методы для настройки и записи логов."""
+
+    @abstractmethod
+    def set_log_strategy(self, writer: Writer):
```
поменяй название метода на set_writer

albrettto replied:
Поменял

vladushka2000 commented on logger.py:
```diff
@@ -0,0 +1,30 @@
+
+    def log(self, level: LogLevel, message: str):
+        """Логирует сообщение с указанным уровнем, используя текущую стратегию записи."""
+        formatted_message = self._format_message(level.name, message)
```
вот тут должен быть как раз не level.name, а level.value

albrettto replied:
Исправил

vladushka2000 commented on writers.py:
```diff
@@ -0,0 +1,41 @@
+        print(message)
+
+
+class FileLogWriter(Writer, FilePathManager):
```
не очень круто наследовать сразу несколько классов, лучше сделать через композицию. Ну ладно, я думаю и так нормально. Поведение таких классов, отнаследованных от нескольких сущностей сложно отслеживать. Особенно когда ты потом делаешь еще один дочерний класс, как UpperCaseFileLogWriter

albrettto committed:
chore: Исправлены недочеты

vladushka2000 commented:
теперь все нравится
